test/Dialect/FIRRTL/layers-errors.mlir

```diff
@@ -0,0 +1,92 @@
+// RUN: circt-opt %s -split-input-file -verify-diagnostics
+
+firrtl.circuit "CastAwayColor" {
+  firrtl.layer @A bind {}
+  firrtl.module @CastAwayColor() {
+    %w = firrtl.wire : !firrtl.probe<uint<1>, @A>
+    // expected-error @below {{cannot discard layer requirements of input ref}}
+    // expected-note  @below {{discarding layer requirements: @A}}
+    %s = firrtl.ref.cast %w : (!firrtl.probe<uint<1>, @A>) -> !firrtl.probe<uint<1>>
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.module @UncoloredDefineUnderLayerblock(out %o : !firrtl.probe<uint<1>>) {
+    %c = firrtl.constant 0 : !firrtl.uint<1>
+    %r = firrtl.ref.send %c : !firrtl.uint<1>
+    firrtl.layerblock @A {
+      // expected-error @below {{'firrtl.ref.define' op has more layer requirements than destination}}
+      // expected-note  @below {{additional layers required: @A}}
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.module @Top(out %o : !firrtl.probe<uint<1>>) {
+    firrtl.layerblock @A {
+      %c = firrtl.constant 0  : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      // expected-error @below {{'firrtl.ref.define' op has more layer requirements than destination}}
+      // expected-note  @below {{additional layers required: @A}}
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.layer @B bind {}
+  firrtl.module @Foo(out %o : !firrtl.probe<uint<1>, @B>) {
+    firrtl.layerblock @B {
+      %c = firrtl.constant 0 : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      %s = firrtl.ref.cast %r : (!firrtl.probe<uint<1>>) -> !firrtl.probe<uint<1>, @B>
+      firrtl.ref.define %o, %s : !firrtl.probe<uint<1>, @B>
+    }
```

Comment:
Nit: can this be removed for test compactness and maintenance?

Further nit: the actual error logic may not differentiate from an instance vs. a wire of the same type. Consider simplifying this. I'm of two opinions of this as this is useful to check, however, it may not actually add coverage to the test. Same thing with the `ref.subop` test.

Reply:
I've changed to an ext module.

My (weakly held) opinion: this is a bit of a behavioural test. We can't test all the things, but it feels reasonable to take wires and instances as a representative subset of our declaration ops, and check that the verifier works for them. We're not just looking for a minimum set of tests that give us coverage based on how the verifier works today.

Reply:
This is a reasonable stance to me. Instances have to work and it's reasonable to lock them in with a test.